File: gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/SurveyConverter.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.persistence.remote.firestore.schema;

import static com.google.android.gnd.model.basemap.BaseMap.typeFromExtension;
import static com.google.android.gnd.util.Localization.getLocalizedMessage;

import com.google.android.gnd.model.Survey;
import com.google.android.gnd.model.basemap.BaseMap;
import com.google.android.gnd.persistence.remote.DataStoreException;
import com.google.common.collect.ImmutableMap;
import com.google.firebase.firestore.DocumentSnapshot;
import java.net.MalformedURLException;
import java.net.URL;
import java8.util.Maps;
import timber.log.Timber;

/** Converts between Firestore documents and {@link Survey} instances. */
class SurveyConverter {

  static Survey toSurvey(DocumentSnapshot doc) throws DataStoreException {
    SurveyDocument pd = doc.toObject(SurveyDocument.class);
    Survey.Builder survey = Survey.newBuilder();
    survey
        .setId(doc.getId())
        .setTitle(getLocalizedMessage(pd.getTitle()))
        .setDescription(getLocalizedMessage(pd.getDescription()));
    if (pd.getJobs() != null) {
      Maps.forEach(pd.getJobs(), (id, obj) -> survey.putJob(JobConverter.toJob(id, obj)));
    }
    survey.setAcl(ImmutableMap.copyOf(pd.getAcl()));
    if (pd.getBaseMaps() != null) {
      convertOfflineBaseMapSources(pd, survey);
    }
    return survey.build();
  }

  private static void convertOfflineBaseMapSources(SurveyDocument pd, Survey.Builder builder) {
    for (BaseMapNestedObject src : pd.getBaseMaps()) {
      if (src.getUrl() == null) {
        Timber.d("Skipping base map source in survey with missing URL");
        continue;
      }
      try {
        URL url = new URL(src.getUrl());
        builder.addBaseMap(
            BaseMap.builder().setUrl(url).setType(typeFromExtension(src.getUrl())).build());
      } catch (MalformedURLException e) {
        Timber.d("Skipping base map source in survey with malformed URL");
      }
    }
  }
}

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SurveyConverter.kt
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.ground.persistence.remote.firestore.schema

import com.google.android.ground.model.Survey
import com.google.android.ground.model.basemap.BaseMap
import com.google.android.ground.model.job.Job
import com.google.android.ground.persistence.remote.DataStoreException
import com.google.android.ground.persistence.remote.firestore.schema.JobConverter.toJob
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableMap
import com.google.firebase.firestore.DocumentSnapshot
import java.net.MalformedURLException
import java.net.URL
import timber.log.Timber

/** Converts between Firestore documents and [Survey] instances. */
internal object SurveyConverter {

  @Throws(DataStoreException::class)
  fun toSurvey(doc: DocumentSnapshot): Survey {
    val pd =
      DataStoreException.checkNotNull(doc.toObject(SurveyDocument::class.java), "surveyDocument")

    val jobMap = ImmutableMap.builder<String, Job>()
    if (pd.jobs != null) {
      pd.jobs.forEach { (id: String, obj: JobNestedObject) -> jobMap.put(id, toJob(id, obj)) }
    }

    val baseMaps = ImmutableList.builder<BaseMap>()
    if (pd.baseMaps != null) {
      convertOfflineBaseMapSources(pd, baseMaps)
    }

    return Survey(
      doc.id,
      pd.title.orEmpty(),
      pd.description.orEmpty(),
      jobMap.build(),
      baseMaps.build(),
      ImmutableMap.copyOf(pd.acl)
    )
  }

  private fun convertOfflineBaseMapSources(
    pd: SurveyDocument,
    builder: ImmutableList.Builder<BaseMap>
  ) {
    for ((url) in pd.baseMaps!!) {
      if (url == null) {
        Timber.d("Skipping base map source in survey with missing URL")
        continue
      }
      try {
        builder.add(BaseMap(URL(url), BaseMap.typeFromExtension(url)))
      } catch (e: MalformedURLException) {
        Timber.d("Skipping base map source in survey with malformed URL")
      }
    }
  }
}
